compiler/rustc_passes/src/hir_stats.rs

```diff
@@ -268,14 +320,62 @@ impl<'v> hir_visit::Visitor<'v> for StatCollector<'v> {
     }
 }
 
+// Used to avoid boilerplate for types with many variants.
+macro_rules! record_variants {
+    (
+        ($self:ident, $val:expr, $kind:expr, $ty:ty, $tykind:ident), // mandatory pieces
+        [$($variant:ident),*],  // non-nullary variants
+        [$($variant0:ident),*]  // nullary variants
+    ) => {
+        match $kind {
+            $(
+                ast::$tykind::$variant(..) => {
+                    $self.record_variant(stringify!($ty), stringify!($variant), Id::None, $val)
+                }
+            )*
+            $(
+                ast::$tykind::$variant0 => {
+                    $self.record_variant(stringify!($ty), stringify!($variant0), Id::None, $val)
+                }
+            )*
```

Comment:
I think `ast::$tykind::$variant {..}` would work for all cases:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=09ab4aa5b8ab740d8ee4a59b721aff8f

Reply:
Nice! I was wondering if there was some trick like that. I'll do it tomorrow.